### PR TITLE
Fix segfaults due to references by adding unsafe_convert methods

### PR DESCRIPTION
**Pull request by musm**
Closes: https://github.com/JuliaInterop/MATLAB.jl/issues/58

### CODE REVIEW

File: src/mxarray.jl [hunk 1]
```diff
@@ -160,10 +165,11 @@ const _mx_is_struct  = mxfunc(:mxIsStruct)
 const _mx_is_cell    = mxfunc(:mxIsCell) 
 
 
-# getting simple attributes
-
 macro mxget_attr(fun, ret)
-    :( ccall($(fun)::Ptr{Void}, $(ret), (Ptr{Void},), mx.ptr) )
+    quote 
+        ptr = unsafe_convert(Ptr{Void}, mx)
```
Comment (yuyichao) [inline src/mxarray.jl]: Don't call `unsafe_convert` directly. Please read the doc and the use case I linked.

Comment (yuyichao) [inline src/mxarray.jl]: Don't call `unsafe_convert` directly -> don't do that for `ccall` at least.

Comment (musm) [inline src/mxarray.jl]: thank you, it's so much better now

File: src/mxarray.jl [hunk 2]
```diff
@@ -159,11 +164,8 @@ const _mx_is_empty   = mxfunc(:mxIsEmpty)
 const _mx_is_struct  = mxfunc(:mxIsStruct)
 const _mx_is_cell    = mxfunc(:mxIsCell) 
 
-
-# getting simple attributes
-
 macro mxget_attr(fun, ret)
-    :( ccall($(fun)::Ptr{Void}, $(ret), (Ptr{Void},), mx.ptr) )
+    :(ccall($(fun)::Ptr{Void}, $(ret), (Ptr{Void},), mx))
```
Comment (yuyichao) [inline src/mxarray.jl]: While you are at it (i.e. for another PR maybe). Note that this is exploiting a lowering bug https://github.com/JuliaLang/julia/issues/4873 so the inputs and any reference of variables from calling scope needs to be escaped. It is also a better practice to NOT implicitly use variables in calling scope and pass them in explicitly instead.

File: src/mxarray.jl [hunk 3]
```diff
@@ -388,12 +390,12 @@ function mxarray{V<:Union{Float64,Bool},I}(a::SparseMatrixCSC{V,I})
     
     mx = mxsparse(V, m, n, nnz)
     
-    ir_p = ccall(_mx_get_ir, Ptr{mwIndex}, (Ptr{Void},), mx.ptr)
-    jc_p = ccall(_mx_get_jc, Ptr{mwIndex}, (Ptr{Void},), mx.ptr)
-    pr_p = ccall(_mx_get_pr, Ptr{V}, (Ptr{Void},), mx.ptr)
+    ir_p = ccall(_mx_get_ir, Ptr{mwIndex}, (Ptr{Void},), mx)
```
Comment (musm) [inline src/mxarray.jl]: These changes probably not necessary, but probably don't hurt.

File: src/mxarray.jl (hunk 2 above)
Comment (musm) [inline src/mxarray.jl]: Thanks, will do. I pushed another commit. I can squash after your approval.

File: src/engine.jl [hunk 4]
```diff
@@ -113,7 +113,7 @@ function put_variable(session::MSession, name::Symbol, v::MxArray)
     @assert libeng::Ptr{Void} != C_NULL
 
     r = ccall(engfunc(:engPutVariable), Cint,
-        (Ptr{Void}, Ptr{UInt8}, Ptr{Void}), session.ptr, string(name), v.ptr)
+        (Ptr{Void}, Ptr{UInt8}, Ptr{Void}), session.ptr, string(name), v)
```
Comment (musm) [inline src/engine.jl]: This change is probably not necessary, but probably don't hurt and are a bit safer.

**Comment by musm** [comment]
Added a few more commits, on my end I am done adding commits (unless changes are requested) to this PR.

Please review/merge as see fit.

**Comment by musm** [comment]
squashed

**Comment by musm** [comment]
Feel free to merge.

There is still a problem with the code (at least no more segfaults), which I will try to address in the future. This is unrelated to these changes, and is related to have default MSessions are handled in the code base, already.

**Comment by musm** [comment]
The example is the following:
```julia
using MATLAB
s = get_default_msession()
close(s)
restart_default_msession()
ERROR: UndefRefError: access to undefined reference
```

A simple try catch around the close in  https://github.com/musm/MATLAB.jl/blob/3e709845c47525c1bf27432221637dc02450f0c2/src/engine.jl#L65   would monkey patch this, but there is probably a better solution to this.

Related, I'm also not too keen on all these methods
`close_default_msession, get_default_msession, restart_default_msession`
to me they seem unnecessary

**Comment by musm** [comment]
fyi I did end up adding a another commit, since I  noticed that `delete` doesn't error on  cleared references. 

Is there a canonical way to  test finalizers.

**Comment by musm** [comment]
This needs another quick review after the last commit,

but I fixed up the issue mentioned earlier here: https://github.com/JuliaInterop/MATLAB.jl/pull/69 (review and merge after this please)

**Pull request merged**